Support precompiled library binaries

eerii opened the pull request:
**This is a work in progress and there are still open questions.**

This PR attempts to support precompiled binaries for libraries. Projects using `system-deps` will be able to specify an url[^1] for downloading and unpacking an archive containing static libraries and `pkg-config` descriptions. Its paths are added to the `PKG_CONFIG_PATH` for each library. This makes it possible to use projects with `-sys` crate dependencies without having to install system dependencies.

## Usage example

Before going into the technical explanation, here is a brief example on how the proposed configuration could work. We have a project `P` that depends on the crate `A`, a safe wrapper for the bindings of `liba`, in crate `A_sys`. Then, `A` depends on `B`, similarly a wrapper around `B_sys`.

```mermaid
flowchart LR
    P --> A
    A --> As{{_A_sys_}}
    A --> B
    B --> Bs{{_B_sys_}}
```

To use prebuilt binaries for `B_sys`, we can add it in the same place we configure the `system_deps` metadata.

<table>
<tr>
<td> P/Cargo.toml </td> <td> A/Cargo.toml </td> <td> B/Cargo.toml </td> <td> B/sys/Cargo.toml </td>
</tr>
<tr>
<td>

```toml
[package]
name = "P"

[dependencies]
A = { ... }
```

</td>
<td>

```toml
[package]
name = "A"

[dependencies]
A_sys = { ... }
B = { ... }
```

</td>
<td>

```toml
[package]
name = "B"

[dependencies]
B_sys = { ... }
```

</td>
<td>

```toml
[package]
name = "B_sys"

[build-dependencies]
system-deps = "7"

[features]
binary = [ "system-deps/binary" ]

[package.metadata.system-deps.B]
name = "libb"
version = "1.0"
url = "https://download/libb.tar.gz"
checksum = "..."
paths = [ "lib/pkgconfig" ]
```

</td>
</tr>
</table>

When building `P`, `libb.tar.gz` will be downloaded. To find `libb`, `system-deps` will instruct `pkg-config` to look first at `DOWNLOAD_DIR/B/lib/pkgconfig`, and if there is a valid library it will be linked. Users of `P` don't have to install `libb`.

Let's say that `A_sys` also has available binaries. The configuration is the same as before. However, imagine that the `A_sys` download already includes a version of `libb` and we want to use that instead. We can make `B_sys` use it like so:

```toml
[package]
name = "A_sys"

[package.metadata.system-deps.A]
name = "liba"
version = "1.0"
url = "https://download/liba.tar.gz"
checksum = "..."
paths = [ "lib/pkgconfig" ]
provides = [ "B" ]
```

This time, `liba.tar.gz` will be downloaded and it will be used to link both `liba` and `libb`. Note that overriding options is only allowed from downstream crates, the outermost configuration will be used.

Finally, if an user building `P` wants to use a local copy of the binaries instead of the provided one, or another url, it can overwrite yet again the configuration set by `A_sys`:

```toml
[package]
name = "P"

[package.metadata.system-deps.A]
url = "file:///path/to/liba"
```

## Previous work

Some crates in the rust ecosystem already provide some similar functionality. Two of the most relevant are [`skia`](https://github.com/rust-skia/rust-skia) and [`openssl`](https://github.com/sfackler/rust-openssl), with different approaches.

- **Skia** uses [`skia-bindings`](https://github.com/rust-skia/rust-skia/tree/master/skia-bindings) to compile the source code in the CI and upload it to a release repository, with different versions for certain feature sets. For users, it first checks if there is a package that matches the version and features requested. If there is none, it falls backs to compiling the code locally.
- **OpenSSL** has a `vendored` feature to control if it should use precompiled binaries. By default it queries the system for the needed libraries, but if `vendored` is enabled it uses `openssl-src` to download and build the source code.

While these are great implementations, they are very specific to the crates they are in and they need certain prerequisites that may not work well for all libraries. For instance, they are working with a base crate with no other `-sys` dependencies that may also want binaries.

## Requisites

As `system-deps` is a general library, the proposal for using precompiled binaries needs to be as **general** as possible. The intended goal is to abstract the difficulty of managing the download and linking but still allow crates to configure when and how their binaries should be used. The only requirement is that they use `pkg-config` files.

It also needs to allow for **multiple crates** in the tree to request and link binaries, since many projects have a combination of libraries using `system-deps`. Additionally, some crates may want to share the same binary archive or may need to override the options of a dependency.

The configuration should be as **easy** as possible for the end user. Ideally, the sys crates should provide the default binary location and the projects using would be able to use it without further configuration. It needs to be possible to specify different downloads for each target and version of the library.

Extra care needs to be taken to avoid unnecessary recompilation or downloads. Using precompiled binaries is specially useful for less powerful systems since it avoid extra compute work, so it is important to optimize the solution.

## Configuration

The main issues when developing this implementation were related to dependency ordering and metadata reading. Cargo is very strict with some of its features, and communicating between crates during compilation is difficult. `system-deps` runs in the build script of the system libraries, once for each dependency that uses it in the project.

### Reuse current system

We need a way of specifying a set of attributes (url, checksum, paths, ...) for each binary archive we want to download. `system-deps` already does something similar with cargo metadata, with one caveat: It only reads the manifest of the crate that it is being compiled. For this use case this is too limiting, since it would mean that every dependency would have to specify its own binary url and that downstream projects would have no way of overriding this behaviour later. It also doesn't allow sharing binary archives across crates, since the downloads would be ran in parallel.

```mermaid
flowchart LR
    P --> A
    A --> As
    As{{_A_sys_}} --> sda[system_deps]
    sda --Read--> As
    A --> B
    B --> Bs
    Bs{{_B_sys_}} --> sdb[system_deps]
    sdb --Read--> Bs
```

### Environment variables

Using environment variables is also not enough. While we could ask the user to set overrides for certain crates (`SYSTEM_DEPS_BINARY_CRATE_NAME_URL`, `SYSTEM_DEPS_BINARY_CRATE_NAME_PKG_PATHS`, ...), this is too verbose, forces the user to add configuration, and it misses one feature: intermediate dependencies overriding their dependencies. Even if we supported adding environment in the build script (#114), we can't do that in middle crates since they are actually running a different instance of `system_deps`. 

```mermaid
flowchart LR
    P --> A
    A --> As
    As{{_A_sys_}} --> sda[system_deps]
    A --> B
    B --> Bs
    Bs{{_B_sys_}} --> sdb[system_deps]
    subgraph .
    Env(**Env**)
    end
    sda & sdb --Read--> .
    As & P -.Can't interact.-x sdb
```

### Patch dependency

A previous prototype tried to "inject" a dependency before `system-deps` ran. This was used as its configuration provider. Since we needed something that ran before _every_ `system_deps` instance, we added a dependency to `system-deps` that could be patched by the caller. This worked because it didn't break the dependency build order, but it was very hacky and it required that users added a `[patch]` pointing to the correct subcrate since that directive has to appear in the top level manifest. Additionally, this didn't allow configuration to come from multiple crates and it was not straightforward to write.

```mermaid
flowchart LR
    P --> A
    A --> As
    As{{_A_sys_}} --> sda[system_deps]
    A --> B
    B --> Bs
    Bs{{_B_sys_}} --> sdb[system_deps]
    
    P -.Patches.-> .
    sda & sdb --Read--> .
    subgraph .
        direction TB
        sdp{{_patched_meta_}} --Replaces--> sdm{{_system_deps_meta_}}
    end
```

### Read metadata from project root

Finally, we arrived at the current solution. If we manage to obtain the top level project directory (the one that has the outmost `Cargo.toml`), we can recursively search the dependency tree from there and read the metadata of its crates.

There is an ongoing discussion on how should cargo provide the **workspace root** of the project (rust-lang/cargo#3946). Other crates such as [`bevy`](https://github.com/bevyengine/bevy) and [`toml-conf`](https://github.com/jamesmunns/toml-cfg) have a similar use case for it as we do. Even if there is no official method to do so at the moment (which we would really appreciate), we found a reliable alternative, inspired by what `toml-conf` [does](https://github.com/jamesmunns/toml-cfg/blob/main/src/lib.rs#L230).

Build scripts get passed one argument that can be read using `std::env::args`. This points to the build directory they have assigned **in the root target directory**. This can be different than reading `OUT_DIR`, which is a target directory in the folder where the dependency is cloned (`/project/dir/target` vs `.cargo/checkout/dep/target`). Because that directory is inside the project, we can run `cargo locate-project` to get the workspace root, which contains the top level `Cargo.toml` manifest. This functionality is implemented in `meta/build::manifest`, more details on the design later.

It works by default in most cases: On different compilation profiles, if the target directory is renamed, using dependencies from crates.io, git and local paths. The only instance in which we found that we can't find it is `TARGET_DIR` is explicitly set to a folder that is not a subdirectory of the top level project (understandably). We provide the `SYSTEM_DEPS_MANIFEST` variable that allows to set the `Cargo.toml` file that will be used. In the rare cases where it fails to locate the project root, it shows the user how to set this variable, which can be as easy as `SYSTEM_DEPS_MANIFEST=$(pwd)/Cargo.toml` or writing `[env] SYSTEM_DEPS_MANIFEST = { value = "Cargo.toml", relative = true }` to `.cargo/config.toml`.

Ideally Rust could provide this path (or at least some way of defining global metadata) so we wouldn't need to do this. However, we feel like this is a good solution that works with minimal user configuration, and the worst case scenario is setting a variable. We could also support alternative ways of providing the binary config, like environment variables as mentioned before.

```mermaid
flowchart LR
    P --> A
    A --> As
    As{{_A_sys_}} --> sda[system_deps]
    A --> B
    B --> Bs
    Bs{{_B_sys_}} --> sdb[system_deps]
    sda & sdb --Read--> sdm{{_system_deps_meta_}}
    sdm -.Finds.-> P
    P -.Read.-> A
    A -.Read.-> B
```

## Design

The final design for the proposed changes is divided in two modules.

### Metadata and binaries

The `system-deps-meta` crate has two purposes:

First, it's build script finds the root manifest as we discussed before. This **needs** to happen in a build script because cargo passes different arguments to libraries and build scripts, and the path we need is only passed for the latter. Additionally, it also exports a target directory (defaults to it's `OUT_DIR`) where the binaries will be stored. This also needs to be defined in the build script because all of the libraries depending on `system_deps` need to have access to the same binary directory to share binaries, so we need to compile it as a constant.

The library part of the crate uses [`cargo-metadata`](https://github.com/oli-obk/cargo_metadata) to parse the root manifest and recursively search the dependency tree for metadata. It returns a table with all of the metadata values related to `system-deps`. The merging function needs some improvements, specially to make sure that only downstream crates can overwrite the previously set metadata, and to handle multiple dependents on the same crate.

The good thing is that this crate can be used from other dependencies to also query the workspace metadata without having to write complex logic. For example, we tested it as a method of specifying which plugins to link in `gstreamer` using the `[package.metadata.gstreamer-plugins]` manifest key, following the same rules as before.

### System deps and pkg config

The `system-deps` build script (in a previous version this was in another separate crate) handles the downloading, unpacking and linking of the binary archives. It almost exclusively happens during its build script for two reasons: The binary download needs to happen **before** any crate that uses `system_deps` gets built, and so that it only happens once.

Finally, in `system-deps`, when we are building each library, we get the paths to the folders that contain the `.pc` files from the downloaded files. We modify the environment in the call to `pkg_config` prepending these paths to the `PKG_CONFIG_PATH` so that the libraries from the binaries are prioritized in the search. The `.pc` files must use a relative prefix so that it works inside of the extracted folder, but this is a common option in meson and other build systems.

### Build process diagram

Since explaining how the build dependencies work can be a mouthful, I think this diagram can be useful to understand which crates are compiled before which. The order is from right to left. Note that the only guarantee is that dependencies are compiled before dependents, and that the build script goes before the library, but the actual order can vary for each run.

```mermaid
flowchart LR
    P --> A
    A --> As
    As{{_A_sys_}} --> sda[system_deps]
    A --> B
    B --> Bs
    Bs{{_B_sys_}} --> sdb[system_deps]
    sda & sdb --> sdbinb
    sdbinb --> sdm
    sdm --> sdmb
    Bs --> sdm

    subgraph download binaries
    sdbinb{{system_deps}}
    end
    subgraph "read
    metadata"
    sdm[meta]
    end
    subgraph find root
    sdmb{{_meta_}}
    end
```

## Motivation

We are working towards adding prebuilt binaries for [`gstreamer-rs`](https://gitlab.freedesktop.org/gstreamer/gstreamer-rs). It has multiple system libraries, and it depends on external ones such as `glib-sys`. Our goal is for the project to provide the official prebuilt binaries so that other crates can use them directly, by only having to enable the `binary` feature (similar to `openssl` using `vendored`). This proposal started as a `gstreamer-rs` only project, and went through quite a bit of iterations, but we realized some of the ideas were generalizable to other crates using `system-deps`.

I still need to rebase, clean and document the `gstreamer-rs` part of the implementation, so it might be messy at the moment, but the development is happening in [this branch](https://gitlab.freedesktop.org/gstreamer/gstreamer-rs).

## TODO:

- [ ] Respect library version keys as `system-deps` does.
- [ ] Add readme and information to `binary` and `meta`.
- [ ] Workshop naming for crates, features and public API.

## Followups

- [ ] Improve caching for reading metadata.
- [ ] Templated URIs.

[^1]: Even if this mentions "download" and "url", local files or folders are allowed using links that start with `file://`. We took inspiration from [skia](https://github.com/rust-skia/rust-skia/tree/master/skia-bindings#prebuilt-binaries-in-an-offline-environment) for this one.

gdesmott commented:
Thanks for this very detailed proposal @eerii !

As [I said](https://gitlab.freedesktop.org/gstreamer/gst-plugins-rs/-/issues/608#note_2645647) I'm all in favor of having this feature as part of `system-deps` but having no experience with this kind of use cases I'd like to hear the input of my fellow GStreamer developers on the design.

cc @nirbheek @thiblahute @xclaesse @sdroege

xclaesse commented:
I'm not a big fan of downloading prebuilt, how do you know it's built for the right platform? OS, arch, libc, etc...

eerii commented:
> I'm not a big fan of downloading prebuilt, how do you know it's built for the right platform

The idea is to allow binaries to be specified using `cfg` target directives, so each platform can download an appropriate archive. I ommited it from the example for brevity.

```toml
[package.metadata.system-deps.A.'cfg(target = "...")']
url = "..."
```

If binaries are requested and there is not a prebuilt archive available for that target, we can always fall back to the default behaviour of `system-deps` and emit a warning.

> From a license POV if that tarball contains (L)GPL code you'll also have to provide script to rebuild it

That would be something that the projects creating the tarballs would have to include in them, not something `system-deps` has to manage, right? Or am I missing something.

Supporting source code downloads is easy. However, I worry that there is too much complexity on the compilation part. Projects have many differences on their build systems and prerequisites, so we would have to find a way of accounting for that. I'm all for adding it as an option, but I think having prebuilts can be very useful in some situations, specially for users with low end hardware that can't afford the compilation time.

xclaesse commented:
> That would be something that the projects creating the tarballs would have to include in them, not something `system-deps` has to manage, right? Or am I missing something.

That's indeed the problem of projects using this feature, not system-deps' problem. Maybe worth mentioning in system-deps doc?

> The idea is to allow binaries to be specified using `cfg` target directives, so each platform can download an appropriate archive.

Makes sense, that can at least help building for major platforms that has a prebuilt provided.

Another question: What's the expected file tree contained inside the tarball? How do you typically generate such tarball? In the GStreamer case I imagine that's going to be a sysroot generated by cerbero? If that's the case, I believe you'll have the problem that .pc files it contains won't be relocatable, that means it will have `-L` flags pointing to `/usr` instead of where you extracted the tarball. It's been a while I haven't touched cerbero, but IIRC it has a tool to rewrite .pc files to relocate them.

xclaesse commented:
BTW, Meson can now generate relocatable pc files, not sure cerbero uses that feature, and it won't work for non-Meson libraries anyway: https://mesonbuild.com/Pkgconfig-module.html#relocatable-pkgconfig-files.

eerii commented:
> Maybe worth mentioning in system-deps doc?

Yes, that is a good idea!

> What's the expected file tree contained inside the tarball?

That's what the `pkg_paths` metadata attribute is for. It is a list of folders inside the tarball in which to look for the `.pc` files. This way we don't enforce any specific structure on them. For the tests I have been doing with gstreamer I have been using this (the main one and the plugins directory):

```toml
pkg_paths = [ "lib/x86_64-linux-gnu/pkgconfig", "lib/x86_64-linux-gnu/gstreamer-1.0/pkgconfig" ]
```

> I believe you'll have the problem that .pc files it contains won't be relocatable

Another thing to make sure to specify in the docs is that this only works with relocatable `.pc` files, they need to point to relative paths inside the binary. I think cerbero already does this since looking at the tarball I built locally they do have relative paths and I didn't change any settings for that:

```pc
# lib/x86_64-linux-gnu/pkgconfig/gstreamer-1.0.pc 
prefix=${pcfiledir}/../../..
datadir=${prefix}/share
includedir=${prefix}/include
libdir=${prefix}/lib/x86_64-linux-gnu
libexecdir=${prefix}/libexec

exec_prefix=${prefix}
toolsdir=${exec_prefix}/bin
pluginsdir=${libdir}/gstreamer-1.0
girdir=${datadir}/gir-1.0
typelibdir=${libdir}/girepository-1.0
pluginscannerdir=${libexecdir}/gstreamer-1.0

Name: gstreamer-1.0
Description: Streaming media framework
Version: 1.25.0.1
Requires: glib-2.0 >=  2.64.0, gobject-2.0, gmodule-no-export-2.0, libunwind
Libs: -L${libdir} -lgstreamer-1.0 -lm -ldl -lnsl
Cflags: -I${includedir}/gstreamer-1.0 -DGST_STATIC_COMPILATION
```

xclaesse commented:
@eerii ok all good if cerbero uses `${pcfiledir}` already, that's what Meson does as well with `-Dpkgconfig:relocatable=true`. That just need to be documented then :)

nirbheek commented:
I finally had time to read through the whole proposal. It's good and seems sound. Some comments:

> Does it make sense to concatenate pkg_paths lists?

Yes. The default should be `@LIBDIR@/pkgconfig`, and the metadata can provide more. Note that @LIBDIR@ can vary per-platform, so it must not be hard-coded in the Cargo.toml, like in the examples provided above.

The way this is usually solved, is that the prebuilt prefix ships with its own pkg-config, which has a value hard-coded in it for where the pc files are. This is obviously wasteful when you're dealing with several different platforms, and might not even work when cross-compiling (the bindir will only contain host binaries, not binaries for your build machine).

There are two possible solutions:

1. Search the prebuilt prefix for `.pc` files and add all such dirs to the list of pkg_paths
2. Require prebuilt prefixes to ship with a `info.toml` file that species this
   - This can be extended for other use-cases later
   
I don't have a strong preference here.

> Does follow or url have priority?

I don't understand this question. "Outermost" values should override "inner" values.

> How to handle two crates overwriting the same dependency with conflicting values?

Throw an error and let the user deal with it. It's always better to be strict, because it's easier to allow use-cases that weren't allowed previously, than disallow use-cases that have always worked incorrectly.

eerii commented:
Thank you so much for the thorough comments!

> The default should be @LIBDIR@/pkgconfig

That makes sense. What I was thinking was using the same cfg directives that handle downloads for different platforms to specify the library paths for each binary:

```toml
[package.metadata.system-deps.A.'cfg(target = "linux_x86...")']
url = "linuxdownload.tar.gz"
pkg_paths = [ "linuxlib/pkg-config" ]

[package.metadata.system-deps.A.'cfg(target = "windows_x86...")']
url = "windowsdownload.tar.gz"
pkg_paths = [ "windowslib/pkg-config" ]
```

But I agree that this can quickly become too verbose and difficult to handle. Searching the prebuilt for `.pc` files sounds good, as long as the packages don't contain conflicting files. I think this is a reasonable expectation, if there are no clear use cases that require otherwise.

> I don't understand this question. "Outermost" values should override "inner" values.

Sorry, I wasn't very clear with this. But yes, the idea is that outermost values override inner ones.

> Throw an error and let the user deal with it. It's always better to be strict, because it's easier to allow use-cases that weren't allowed previously, than disallow use-cases that have always worked incorrectly.

And yes, I have been working on ensuring that this is done correctly and analyzing the different package graph cases and which are conflicting. I'm almost done fixing the implementation to account for this, I'll upload the revised version later today.

nirbheek commented:
> I think I prefer your approach of requiring the user to specify the paths instead of trying to guess tbh, we should also probably show that right from the start in the documentation.

I think that's a property of the prefix (and hence the tarball). It doesn't make sense to put in the configuration, and it just makes it easy to get it wrong. That's why I suggested that the prefix/tarball could specify it in a `config.toml` or something, if we don't want to do autodetection.

nirbheek commented:
For example, if you want to add support for a new target (say, something based on RISC-V, or NixOS), you don't need to update every -sys crate (and potentially every project that uses that -sys crate). You just build some new binary tarballs and everything works.

thiblahute commented:
> For example, if you want to add support for a new target (say, something based on RISC-V, or NixOS), you don't need to update every -sys crate (and potentially every project that uses that -sys crate). You just build some new binary tarballs and everything works.

You will have to anyway to add the URI to the new tarball anyway, unless we add some templating logic, which could also make sense.

> > I think I prefer your approach of requiring the user to specify the paths instead of trying to guess tbh, we should also probably show that right from the start in the documentation.
> 
> I think that's a property of the prefix (and hence the tarball). It doesn't make sense to put in the configuration, and it just makes it easy to get it wrong. That's why I suggested that the prefix/tarball could specify it in a `config.toml` or something, if we don't want to do autodetection.

That approach is fine by me, I do not like autodetection. Also it could be added later fmopv as we already have a working solution.

Also, what is the point of having the .pc files in the tarballs platform dependent? They are never meant to be used on  a system wide prefix, but should be consumed by intermediary apps. I have the impression that we could default to `lib/pkgconfig` and make it all simpler no?

nirbheek commented:
> You will have to anyway to add the URI to the new tarball, unless we add some templating logic, which could also make sense.

I thought that was the plan already, the manylinux URIs look templated :)

> Also, what is the point of having the .pc files in the tarballs platform dependent? They are never meant to be used on a system wide prefix, but should be consumed by intermediary apps. I have the impression that we could default to lib/pkgconfig and make it all simpler no?

That's how distros work. Libraries and executables built for a particular distro will look for that distro's' libdir in whatever prefix you use. We tried to not do this in cerbero already, and that failed.

thiblahute commented:
> > Also, what is the point of having the .pc files in the tarballs platform dependent? They are never meant to be used on a system wide prefix, but should be consumed by intermediary apps. I have the impression that we could default to lib/pkgconfig and make it all simpler no?
> 
> That's how distros work. Libraries and executables built for a particular distro will look for that distro's' libdir in whatever prefix you use. We tried to not do this in cerbero already, and that failed.

Interesting, how did that fail?

nirbheek commented:
> Interesting, how did that fail?

I don't remember anymore. Something around how modules are searched. All I remember is that in cerbero we tried to enforce `prefix/lib` everywhere and had to revert it.

eerii commented:
Hi! Update on this, I have been refactoring the proposal, cleaning old changes and improving the interface. I think that it is now at a point where I'm pretty happy with the public API and how the parsing system works. There may be a few details to still work on, but I think it is a good place to start reviewing this.

I improved the error handling and added tests for both the metadata parsing and the binary downloads and filetypes. Since there are complex dependency cases with multiple pacakges, each new test generates the needed `Cargo.toml` files. This makes it easier to write tests and check a few variations on each.

I updated the PR description, but here is a list of the important changes:

#### Improvements on metadata parsing

The naive previous implementation only worked for very simple package trees and it didn't fail properly on conflicts and sometimes it produced wrong results. I have reworked this part from scratch and it should hopefully be more robust and able to handle more generic cases. It is also reusable outside of `system-deps` (for example, to read `gstreamer-rs` plugins).

First, the base package(s) is read from the root manifest. For each one, we read the metadata and descend into its (unvisited) dependencies, building an "upside down" dependency tree containing the metadata. Starting from the leaves we reduce the value applying the correct overwrites and checking for conflicts. The final result is a single map with the calculated values.

It also has many little improvements, such as type checking for values, extending lists instead of overwriting them, using `cfg` to gate configurations behind a target and more. Additionally, as @nirbheek suggested, paths can be read from a `info.toml` at the base level of the downloaded archive:

```toml
# info.toml
paths = [ "lib/pkgconfig" ]
```

#### Reworked how prebuilt binaries are shared:
 
The `follows` key has been changed to `provides`. Instead of placing it on each package that wants to reuse binaries from another, the original package can have a list of what libraries are contained in the binary it is downloading. This is more concise and avoids having to modify other manifests.

```toml
# New version
[package.metadata.system-deps.gstreamer_1_0]
url = "..."
checksum = "..."
provides = [ "glib_2_0", "gobject_2_0", "gio_2_0", "gst*" ]

# Before
[package.metadata.system-deps.gstreamer_1_0]
url = "..."
checksum = "..."

[package.metadata.system-deps.glib_2_0]
follows = "gstreamer_1_0"

[package.metadata.system-deps.gobject_2_0]
follows = "gstreamer_1_0"

# Many more lines ...
```

Internally, it is actually adding the `follows` key to the packages when building the final metadata tree (this is done to reuse the same overwriting and merging rules and to ensure that there are no conflicts), so the previous version still technically works.

Additionally, **wilcards** are supported for provided packages. This is particularly useful for `gstreamer` where there are too many libraries to reliably list. It is important to note that wildcards are a bit special and they never overwrite a previous package url, as opposed to writing the full name in `provides`. There are a few technical reasons why this was kept like that, mainly code complexity and performance overhead in metadata resolution. I think they are useful but we should clearly explain how they work in the documentation.

#### Downloading and decompressing

These actions should now be faster and more reliable. They now happen in parallel and only if there wasn't a valid version. Checksums are properly calculated and checked.

The library for handling the downloads has been changed from [`reqwest`](https://crates.io/crates/reqwest) to [`attohttpc`](https://crates.io/crates/attohttpc), since it is more lightweight and it doesn't depend on an async runtime.

#### Reorganize the proposal structure

Instead of having three crates and everything all over the place, now the logic is on the `system-deps-meta` crate (gated by features) so that it can be easily reused by other projects.

- **Build** `system-deps-meta`: Finds project root.
- **Lib** `system-deps-meta`: Utils for parsing metadata and downloading binaries.
- **Build** `system-deps`: Binaries are downloaded and paths are serialized (before any dependent crate is built).
- **Lib** `system-deps`: Probe libraries, set pkg-config paths, ...

#### What's next?

The current version now finally covers the use cases indicated in the proposal, works for projects with complex dependency trees, and it has tests and error handling. There are a few details missing that are important before merging:

- Although it is somewhat complete, finish the bits of documentation left and add an example to the README.
- `system-deps` allows to specify multiple versions of a package with `[system-deps.name.version]`. At the moment the `url` can only be specified at the top level, but adding version support shouldn't be too hard.
- Up until now I have been calling these feature "binary" but I'm not too sold on the name. Other options are "prebuilt", "precompiled", "download", "vendored". What do you think we should refer to this feature as? I think it is a good idea to do a pass rewriting the docs and naming of everything to be more cohesive.

There are other features that I think are more suited to a followup PR, since this is already big enough and they are not essential:

- Unify `system-deps-meta` and `metadata.rs`. This is not hard, we could probably do it now, but it will be tricky to do without adding dependencies. I was more lenient on `system-deps-meta` since it is opt-in and it will not affect existing users, but `cargo-metadata` uses `serde-json` so we are duplicating the serialization libraries. So I think this will require more careful planning.
- Templated URIs and other similar features.
- Caching is not a big issue since `read_metadata` is only called once on the build script, which all crates then share. However, I'm sure some improvements can be made to make this even more performant.

gdesmott commented:
Thanks for all your work on this!

I'll let @thiblahute and @nirbheek review the design and user API as they know much more than me on this use case.

Btw, feel free to clean up the branch (squash commits, etc) as it will be required before merging.

gdesmott commented:
Should we go forward with this?

@nirbheek @thiblahute : I think you were interested about such feature, right?

thiblahute reviewed (commented):
I had started to review that a while back and only had those comments, I wanted to have another deeper look but I don't have time right now.

I would love to see that MR finished/merged yes

eerii commented on build.rs:
```diff
@@ -0,0 +1,32 @@
+        // Write the binary paths to a file for later use
+        let dest = Path::new(BUILD_TARGET_DIR).join("paths.toml");
+        fs::write(&dest, paths.to_string()?).map_err(BinaryError::InvalidDirectory)?;
+        println!("cargo:rustc-env=BINARY_PATHS={}", dest.display());
```
That's a good idea, we can prefix this variables with `SYSTEM_DEPS_` to avoid any clashes.

eerii commented:
I also had some cleanup changes and fixes from a while ago, let me review those and I'll update the branch c: Would also love to see this merged.

Apologies for not updating this sooner, I had to deal with some issues but everything is better now.

eerii commented on meta/src/binary.rs:
```diff
@@ -0,0 +1,487 @@
+}
+
+/// Iteratively scan `dir` for subdirectories named "pkgconfig" and return their paths.
+fn find_pkgconfig_dirs(dir: &Path) -> Vec<PathBuf> {
```
Should this have a recursion limit to avoid hangs on very deep binaries?

Also, it might be good to avoid traversing symlinks. I don't know who would add a symlink to a binary tarball, but if there is one this could loop forever ([`is_dir`](https://doc.rust-lang.org/std/path/struct.Path.html#method.is_dir) follows symlinks).

eerii reviewed (commented):
Thanks :) I left some comments and questions to discuss. I would also like to rename some variables since I did a poor name with that (`l`, `n`, `c`...), but that's not as important now.

calvaris commented:
In this new version I added support for Mac and Windows and addressed the comments, at least most of them.

eerii commented on build.rs:
```diff
@@ -0,0 +1,59 @@
+pub fn main() {
+    #[cfg(feature = "binary")]
+    binary::build().unwrap_or_else(|e| panic!("{}", e));
+}
```
```suggestion
pub fn main() -> Result<(), Box<dyn std::error::Error>> {
    #[cfg(feature = "binary")]
    binary::build()?;
    Ok(())
}
```

eerii commented on build.rs:
```diff
@@ -0,0 +1,59 @@
+                .map(|d| d.display().to_string())
+                .collect();
+            if !dirs.is_empty() {
+                let joined = dirs.join(";");
```
Maybe it's better to use [`std::env::join_paths`](https://doc.rust-lang.org/std/env/fn.join_paths.html)? I think just joining them could fail in some cases.

eerii commented on src/metadata.rs:
```diff
@@ -339,6 +344,12 @@ impl MetaData {
 
                     dep.version_overrides.push(builder.build()?);
                 }
+                ("url", DeValue::String(_)) => {}
```
```suggestion
				// Fields used by system-deps-meta
                ("url", DeValue::String(_)) => {}
```

eerii commented on meta/src/error.rs:
```diff
@@ -0,0 +1,135 @@
+    /// The cfg() expression is valid, but not currently supported.
+    UnsupportedCfg(String),
+}
+
```
```suggestion

impl std::error::Error for Error {}

```

eerii commented on meta/src/error.rs:
```diff
@@ -0,0 +1,135 @@
+                write!(f, "Unsupported cfg() expression: {}", s)
+            }
+            #[cfg(feature = "binary")]
+            Self::Binary(e) => e.fmt(f),
```
```suggestion
            Self::CfgNotObject(expr) => {
                write!(f, "The expression '{}' is not guarding a package", expr)
            }
            Self::DeserializeError(e) => write!(f, "Error while parsing: {}", e),
            Self::IncompatibleMerge => write!(f, "Can't merge metadata"),
            Self::PackageNotFound(pkg) => write!(f, "Package not found: {}", pkg),
            Self::SerializeError(e) => write!(f, "Error while parsing: {}", e),
            Self::InvalidCfg(e) => write!(f, "Invalid cfg() expression: {}", e),
            Self::UnsupportedCfg(expr) => {
                write!(f, "Unsupported cfg() expression: {}", expr)
            }
            #[cfg(feature = "binary")]
            Self::Binary(e) => e.fmt(f),
```

eerii commented on meta/src/error.rs:
```diff
@@ -0,0 +1,135 @@
+    impl From<attohttpc::Error> for BinaryError {
+        fn from(e: attohttpc::Error) -> Self {
+            Self::DownloadError(e)
+        }
+    }
+
+    impl fmt::Display for BinaryError {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            match self {
+                Self::DecompressError(e) => {
+                    write!(f, "Failed to decompress the binary archive: {}", e)
+                }
+                Self::DirectoryIsFile(s) => {
+                    write!(f, "The binary target directory is a file: {}", s)
+                }
+                Self::DownloadError(e) => write!(f, "Failed to download binary archive: {}", e),
+                Self::InvalidChecksum(p, a, b) => {
+                    write!(
+                        f,
+                        "Mismatch in the checksum of {}:\n\
+                        - Specified: {}\n\
+                        - Calculated: {}",
+                        p, a, b
+                    )
+                }
+                Self::InvalidDirectory(e) => {
+                    write!(f, "The binary target directory is not valid: {}", e)
+                }
+                Self::InvalidFollows(a, b) => {
+                    write!(f, "The package {} follows {}, which doesn't exist", a, b)
+                }
+                Self::LocalFileError(e) => {
+                    write!(f, "The requested local folder could not be read: {}", e)
+                }
+                Self::SymlinkError(e) => {
+                    write!(f, "Couldn't create symlink to local binary folder: {}", e)
+                }
+                Self::UnsupportedExtension(s) => {
+                    write!(f, "Unsupported binary extension for {}", s)
+                }
```
```suggestion
                Self::DecompressError(e) => {
                    write!(f, "Failed to decompress the binary archive: {}", e)
                }
                Self::DirectoryIsFile(path) => {
                    write!(f, "The binary target directory is a file: {}", path)
                }
                Self::DownloadError(e) => write!(f, "Failed to download binary archive: {}", e),
                Self::InvalidChecksum(pkg, specified, calculated) => {
                    write!(
                        f,
                        "Mismatch in the checksum of {}:\n\
                        - Specified: {}\n\
                        - Calculated: {}",
                        pkg, specified, calculated
                    )
                }
                Self::InvalidDirectory(e) => {
                    write!(f, "The binary target directory is not valid: {}", e)
                }
                Self::InvalidFollows(pkg, follows) => {
                    write!(f, "The package {} follows {}, which doesn't exist", pkg, follows)
                }
                Self::LocalFileError(e) => {
                    write!(f, "The requested local folder could not be read: {}", e)
                }
                Self::SymlinkError(e) => {
                    write!(f, "Couldn't create symlink to local binary folder: {}", e)
                }
                Self::UnsupportedExtension(ext) => {
                    write!(f, "Unsupported binary extension: {}", ext)
                }
```

eerii commented on meta/build.rs:
```diff
@@ -0,0 +1,105 @@
+
+/// Try to find the project root using locate-project
+#[cfg(not(windows))]
+fn find_with_cargo(dir: &Path) -> Option<PathBuf> {
```
Hmm interesting, I didn't know about the deadlock behaviour on Windows. Let's call this `find_workspace_root` as well, they are mutually exclusive and that way we don't have to do the `cfg` below.

```suggestion
fn find_workspace_root(dir: &Path) -> Option<PathBuf> {
```

eerii commented on meta/build.rs:
```diff
@@ -0,0 +1,105 @@
+    #[cfg(not(windows))]
+    let result = find_with_cargo(&dir);
+    #[cfg(windows)]
+    let result = find_workspace_root(&dir);
```
Then this can just be:

```suggestion
    let result = find_workspace_root(&dir);
```

eerii commented on src/test_binary.rs:
```diff
@@ -0,0 +1,621 @@
+}
+
+// TODO: Library versions test
+
```
```suggestion
```

There is already a test below, even if it is disabled.

eerii commented on src/lib.rs:
````diff
@@ -202,27 +202,94 @@
+//! ```
+//!
+//! TODO: The environment variables `SYSTEM_DEPS_BINARY_URL` and `SYSTEM_DEPS_BINARY_CHECKSUM` will allow setting a global
+//! binary url (for example, for local testing).
````
Do we still needd this?

eerii commented on src/lib.rs:
```diff
@@ -1120,43 +1265,55 @@ impl Library {
+impl<T: AsRef<Path>> PathOrList<T> for T {
+    fn join_paths(&self, other: impl AsRef<[PathBuf]>) -> OsString {
+        let other = other.as_ref().iter().map(|p| p.as_path());
+        env::join_paths(iter::once(self.as_ref()).chain(other)).unwrap()
```
```suggestion
        env::join_paths(iter::once(self.as_ref()).chain(other))
            .expect("Path contains invalid character")
```

eerii commented on src/lib.rs:
```diff
@@ -1120,43 +1265,55 @@ impl Library {
+    fn join_paths(&self, other: impl AsRef<[PathBuf]>) -> OsString {
+        let slice: &[T] = (*self).borrow();
+        let other = other.as_ref().iter().map(|p| p.as_path());
+        env::join_paths(slice.iter().map(|p| p.as_ref()).chain(other)).unwrap()
```
```suggestion
        env::join_paths(slice.iter().map(|p| p.as_ref()).chain(other))
            .expect("Path contains invalid character")
```

eerii commented on src/lib.rs:
```diff
@@ -1120,43 +1265,55 @@ impl Library {
+    fn join_paths(&self, other: impl AsRef<[PathBuf]>) -> OsString {
+        match self {
+            Some(s) => s.join_paths(other),
+            None => env::join_paths(other.as_ref().iter().map(|p| p.as_os_str())).unwrap(),
```
```suggestion
            None => env::join_paths(other.as_ref().iter().map(|p| p.as_os_str()))
                .expect("Path contains invalid character"),
```

eerii commented on meta/src/parse.rs:
```diff
@@ -0,0 +1,489 @@
+/// On Windows, `cargo metadata` deadlocks when called from a build script
+/// because cargo holds mandatory file locks on the workspace.
+#[cfg(windows)]
+fn collect_nodes(manifest: &Path, section: &str) -> Result<HashMap<String, MetadataNode>, Error> {
```
Is there an issue for this and the `locate-project` behaviour on Windows that we can link?

Since we already have to rebuild the metadata parser we might as well remove the `cargo_metadata` dependency completely. But I'd say let's do that in a followup patch. The metadata parser was always intended to be temporary and replaced by an extension of the `src/metadata.rs` that we already had. I think it will be *much* easier to do that refactor once this huge PR is merged.

calvaris replied:
Ok, let's do it in a follow up then.

eerii commented on meta/src/binary.rs:
```diff
@@ -0,0 +1,942 @@
+            e => Err(BinaryError::UnsupportedExtension(
+                e.to_str().unwrap().into(),
+            )),
+        }
```
```suggestion
        let extension = path
           .extension()
           .and_then(|s| s.to_str())
           .unwrap_or("<error>");
        match extension {
            #[cfg(feature = "pkg")]
            "pkg" => Ok(Extension::Pkg),
            #[cfg(feature = "nsis")]
            "exe" => Ok(Extension::Nsis),
            #[cfg(feature = "gz")]
            "gz" | "tgz" => Ok(Extension::TarGz),
            #[cfg(feature = "xz")]
            "xz" => Ok(Extension::TarXz),
            #[cfg(feature = "zip")]
            "zip" => Ok(Extension::Zip),
            e => Err(BinaryError::UnsupportedExtension(e.into())),
        }
```

thiblahute replied:
``` 
-            .unwrap_or("<error>");
+           .unwrap_or("<none>");
```

calvaris replied:
Will be solved in next push

eerii commented on meta/src/binary.rs:
```diff
@@ -0,0 +1,942 @@
+        for entry in entries.flatten() {
+            let path = entry.path();
+            // Use symlink_metadata (lstat) to avoid following symlinks
+            let Ok(meta) = entry.metadata() else {
```
```suggestion
            let Ok(meta) = entry.symlink_metadata() else {
```

I'm not sure if this is fixing the symlink issue. [`is_dir`](https://doc.rust-lang.org/std/fs/struct.Metadata.html#method.is_dir) says it will be false for symlink metadata obtained from [symlink_metadata](https://doc.rust-lang.org/std/fs/fn.symlink_metadata.html), but we aren't using that function. I think we may want to call that instead.

eerii commented on meta/src/binary.rs:
```diff
@@ -0,0 +1,942 @@
+
+    let mut result = Vec::new();
+    let mut queue = vec![(dir.to_path_buf(), 0usize)];
+    while let Some((current, depth)) = queue.pop() {
```
```suggestion
    let mut stack = vec![(dir.to_path_buf(), 0usize)];
    while let Some((current, depth)) = stack.pop() {
```

Small naming nit, but this is LIFO, not FIFO

eerii commented on meta/src/binary.rs:
```diff
@@ -0,0 +1,942 @@
+                if path.file_name().is_some_and(|n| n == "pkgconfig") {
+                    result.push(path);
+                } else if depth < MAX_DEPTH {
+                    queue.push((path, depth + 1));
```
```suggestion
                    stack.push((path, depth + 1));
```

eerii reviewed (commented):
Looks good! :)

thiblahute commented on meta/Cargo.toml:
```diff
@@ -0,0 +1,31 @@
+xz = { version = "0.1", optional = true }
+tar = { version = "0.4", optional = true }
+zip = { version = "2.4", optional = true }
+pkg-extractor = { git = "https://github.com/thiblahute/pkg_extractor", optional = true }
```
You should use the release I just published here: https://crates.io/crates/pkg-extractor/0.1.1 now :-)

calvaris replied:
This is left as it is because of the lzma rework.

thiblahute commented on meta/src/binary.rs:
```diff
@@ -0,0 +1,942 @@
+    /// might happen more often. It has no effect if `url` is a local folder.
+    checksum: Option<String>,
+    /// A list of relative paths inside the binary archive that point to a folder containing
+    /// package config files. These directories will be prepended to the `PKG_CONFIG_PATH` when
```
```suggestion
    /// package config files (`.pc`). These directories will be prepended to the `PKG_CONFIG_PATH` when
```

calvaris replied:
Will be solved in next push

thiblahute commented on meta/src/binary.rs:
```diff
@@ -0,0 +1,942 @@
+    false
+}
+
+/// Search for an existing GStreamer NSIS installation on Windows.
```
That whole GStreamer specific logic doesn't belong here at all, we need to think about it again, but that whole nsis feature seems weird to me and I have the impression we should simply extract the .exe ourself in the right place and use that instead.

calvaris replied:
We moved to inno. This should be done.

thiblahute reviewed (changes requested):
I didn't review it all, but I think it looks good apart from the windows nsis feature we need to clean up.

calvaris commented on meta/src/binary.rs:
```diff
@@ -0,0 +1,487 @@
+
+                // Allow multiple downloads at the same time
+                if !valid {
+                    s.spawn(move || make_available(bin, &dst).map_err(|e| panic!("{}", e)));
```
I wouldn't fallback to the system. It's quite misleading.

thiblahute commented on meta/src/binary.rs:
```diff
@@ -0,0 +1,631 @@
+    {
+        // mklink requires backslash paths; forward slashes cause
+        // "invalid modifier" errors in cmd.exe.
+        let dst_str = dst.to_string_lossy().replace('/', "\\");
```
Isn't there a better API for that?

calvaris commented on meta/src/lib.rs:
```diff
@@ -0,0 +1,17 @@
+pub const BUILD_MANIFEST: &str = env!("BUILD_MANIFEST");
+
+/// Directory where `system-deps` related build products will be stored.
+pub const BUILD_TARGET_DIR: &str = env!("BUILD_TARGET_DIR");
```
We have a TODO for this.

nirbheek commented:
@thiblahute I would **strongly** recommend using a pkg-config in the downloaded prefix if available over anything on the system.

1. People often have random broken pkg-config implementations scattered around in their PATH on Windows
2. pkg-config can be built with options that can break things, for example by inserting "system" includedirs and libdirs, or for auto-detecting the prefix

So there is no guarantee that you can consume the gstreamer binaries with any pkg-config other than what we ship in the installer / tarball.

jschwe commented:
> Build scripts get passed one argument that can be read using std::env::args. This points to the build directory they have assigned in the root target directory. This can be different than reading OUT_DIR, which is a target directory in the folder where the dependency is cloned (/project/dir/target vs .cargo/checkout/dep/target). Because that directory is inside the project, we can run cargo locate-project to get the workspace root, which contains the top level Cargo.toml manifest. 

Note that users can set [`CARGO_TARGET_DIR`](https://doc.rust-lang.org/cargo/reference/config.html#buildtarget-dir) to an arbitrary directory, which would break this approach. Having a global target dir is apparently not that uncommon, to cache compilation results (see also next link).
How does this interact with the [Cargo cross workspace cache project goal for 2026](https://rust-lang.github.io/rust-project-goals/2026/cargo-cross-workspace-cache.html)? 

On a different note, I'm curious if this proposal would also work for prebuilt shared libraries? Rebuild times can be dominated by link times, and having a shared library would help reduce that (at the cost that the user needs some additional configuration on their side, but for many projects that might be a tradeoff worth making for larger libraries).

nirbheek commented:
> On a different note, I'm curious if this proposal would also work for prebuilt shared libraries

The primary target here is prebuilt shared libraries, which is the only thing supported by GStreamer via Cerbero on Windows (for now — you can always build gstreamer with static libs yourself from source), and the preferred configuration on macOS and Linux.

eerii commented:
> Note that users can set [`CARGO_TARGET_DIR`](https://doc.rust-lang.org/cargo/reference/config.html#buildtarget-dir) to an arbitrary directory, which would break this approach. Having a global target dir is apparently not that uncommon, to cache compilation results (see also next link).

Yes, this is a limitation. Unfortunately I don't think there is a better way of reliably finding the main `Cargo.toml` yet. So as of now this is a best effort approach. See https://github.com/rust-lang/cargo/issues/3946.

As a workaround for those tricky cases, users can set the `SYSTEM_DEPS_BUILD_MANIFEST` var to point to it. If it can't be found, we print a warning [here](https://github.com/gdesmott/system-deps/pull/116/changes#diff-4b956f1bb16a0d8cc6f4003775399febb100497f08cd7a82e722825f2dfb44d2R79) to let users know what to do:

> Error determining the cargo root manifest.
> Please set `SYSTEM_DEPS_BUILD_MANIFEST` to the path of your project's Cargo.toml.

calvaris committed:
binary: support prebuilt library binaries

Add a new `binary` feature that lets crates declare prebuilt system
library archives in `[package.metadata.system-deps]`, which are
downloaded, extracted and added to `PKG_CONFIG_PATH` automatically at
build time. Metadata is collected across the whole crate graph (not
just the current package), so dependents can override or extend the
binaries provided by their dependencies.

Highlights:

* New `system-deps-meta` companion crate that walks the cargo metadata,
  reduces per-package metadata tables with proper inheritance, and
  evaluates `cfg()` expressions recursively (including inside dependency
  tables).
* Archive download + extraction with optional `gz`, `xz` and `zip`
  features; URL templates and per-target `url`/`checksum` entries;
  `follows` to share a single archive between several libraries.
* Automatic discovery of `pkgconfig` directories inside extracted
  archives, so users don't need to spell out `pkg_paths` in the common
  case.
* `pkg` feature (macOS only): extract Apple flat packages (.pkg)
  end-to-end, including PBZX payloads, via `pkg-extractor`.
* `inno` feature (Windows only): extract Inno Setup `.exe` installers
  via the `extract` feature of the `inno` crate.

Two cfg-gated MSVC workarounds bundled: extracted `.pc` files are
rewritten to bypass pkg-config 0.29.2's broken auto-prefix heuristic,
and the `static=` link modifier is skipped to dodge rustc's
`ar_archive_writer` u32 overflow on huge gst-plugins-rs archives
(https://github.com/rust-lang/ar_archive_writer/issues/31).

Co-Authored-By: Thibault Saunier <tsaunier@igalia.com>
Co-Authored-By: Eri Pazos <epazos@igalia.com>

thiblahute commented on meta/src/binary.rs:
```diff
@@ -0,0 +1,790 @@
+/// `pkgconf` (the modern replacement for the original `pkg-config`) does not
+/// suffer from this. We keep this workaround because the GStreamer Windows
+/// installer still ships `pkg-config 0.29.2`.
+#[cfg(windows)]
```
Can't we move that workaround to gstreamer-rs build.rs file directly by any chance?

calvaris replied:
I'd rather keep it here. Extraction is owned by system-deps-meta, not by gstreamer-rs: the rewrite has to run inside make_available, after decompression and before any sys-crate's build.rs probes the bundle with pkg-config — and cargo gives us no deterministic ordering between sibling sys-crates, so moving the patch into gstreamer-rs would force every one of them (gstreamer-sys, gstreamer-base-sys, gstreamer-video-sys, …) to synchronise on it. The bug isn't GStreamer-specific either: any bundle distributed via the binary feature whose .pc files live in nested subdirectories (lib/<something>/pkgconfig/foo.pc) trips pkg-config 0.29.2's --define-prefix heuristic on Windows, so keeping the fix here avoids every future consumer of system-deps binary having to re-implement it. It also keeps the rewrite where it semantically belongs — mutating files under <TARGET_DIR>/<name>/, a directory wholly owned by system-deps-meta and not visible to gstreamer-rs except via the public Paths API. Finally, the doc comment already flags this as a self-deprecating workaround: once the GStreamer Windows installer ships pkgconf (the modern fork without this bug) instead of pkg-config 0.29.2, the whole function goes away.

nirbheek replied:
We build our own pkg-config in cerbero, so we can fix the bug in pkg-config too: https://gitlab.freedesktop.org/gstreamer/meson-ports/pkg-config

nirbheek replied:
Actually, another thing: we're not supposed to use --define-prefix with pkg-config anymore in Cerbero. All pc files have a relative prefix defined: `prefix=${pcfiledir}/../..` or `prefix=${pcfiledir}/../../..` (for plugins). I am surprised that you are running into this bug at all.

nirbheek replied:
https://gitlab.freedesktop.org/gstreamer/cerbero/-/merge_requests/2257

calvaris replied:
Nice, I guess we can wait and once this is done I'll update the PR. I just uploaded a version still with this workaround.

nirbheek replied:
It has already been merged and backported. It will be in 1.28.4, so you could wait for that perhaps.

thiblahute replied:
Is there a build @calvaris can try with?

calvaris replied:
If someone presses play and generate the artifacts in the MR, yes.

calvaris replied:
@nirbheek , just filed https://gitlab.freedesktop.org/gstreamer/cerbero/-/merge_requests/2264 to solve a linking issue in servo

calvaris replied:
As we discussed on Matrix, I did the changes you asked for. All seems to check out including servo.

eerii reviewed (commented):
Looks good to me too. Would we need a `pkg-extractor` release too?

gdesmott commented on meta/Cargo.toml:
```diff
@@ -0,0 +1,38 @@
+edition.workspace = true
+documentation.workspace = true
+
+[dependencies]
```
Can we centralize all the deps in the workspace `Cargo.toml`?
I find it more convenient.

calvaris committed:
binary: support prebuilt library binaries

Add a new `binary` feature that lets crates declare prebuilt system
library archives in `[package.metadata.system-deps]`, which are
downloaded, extracted and added to `PKG_CONFIG_PATH` automatically at
build time. Metadata is collected across the whole crate graph (not
just the current package), so dependents can override or extend the
binaries provided by their dependencies.

Highlights:

* New `system-deps-meta` companion crate that walks the cargo metadata,
  reduces per-package metadata tables with proper inheritance, and
  evaluates `cfg()` expressions recursively (including inside dependency
  tables).
* Archive download + extraction with optional `gz`, `xz` and `zip`
  features; URL templates and per-target `url`/`checksum` entries;
  `follows` to share a single archive between several libraries. Local
  `file://` archives and plain directories are supported too (a
  directory is linked with a junction on Windows, a symlink elsewhere).
* Automatic discovery of `pkgconfig` directories inside extracted
  archives, recorded in a generated `info.toml`, so users don't need to
  spell out `pkg_paths` in the common case.
* When a bundle ships its own `bin/pkg-config`, it is preferred over the
  host one (unless `PKG_CONFIG` is set explicitly) so that probing
  matches the bundled libraries.
* `pkg` feature (macOS only): extract Apple flat packages (.pkg)
  end-to-end, including PBZX payloads, via `pkg-extractor`.
* `inno` feature (Windows only): extract Inno Setup `.exe` installers
  via the `extract` feature of the `inno` crate.
